File: cpp/src/internal/TokenBuilder.cpp
#include "internal/TokenBuilder.h"
#include "internal/Token.h"
#include <string>

namespace {
  struct StringTokenBuilder : TokenBuilder {
    void append(char c) override {
      str += c;
    }

    Token* build() override {
      return Token::newStringToken(str);
    }

    std::string str;
  };

  struct NumericTokenBuilder : TokenBuilder {
    NumericTokenBuilder() : length(0), value(0) {
    }

  private:
    void append(char c) override {
      length++;
      value = 10 * value + intValue(c);
    }

    Token* build() override {
      return Token::newNumericToken(length, value);
    }

  private:
    static int intValue(char c) {
      return std::isdigit(c) ? (c - '0') : c;
    }

  private:
    int length;
    int value;
  };
}

TokenBuilder* TokenBuilder::newStringTokenBuilder() {
  return new StringTokenBuilder();
}

TokenBuilder* TokenBuilder::newNumericTokenBuilder() {
  return new NumericTokenBuilder();
}

File: java/src/main/java/coach2018/TokenBuilder.java
package coach2018;

interface TokenBuilder {
    void  append(char c);
    Token build();
}

class StringTokenBuilder implements TokenBuilder {
    private StringBuilder buff = new StringBuilder();

    @Override
    public void append(char c) {
        buff.append(c);
    }

    @Override
    public Token build() {
        return new StringToken(buff.toString());
    }
}

class NumericTokenBuilder implements TokenBuilder {
    private int length;
    private int value;

    @Override
    public void append(char c) {
        length++;
        value = 10 * value + intValue(c);
    }

    private static int intValue(char c) {
        return Character.isDigit(c) ? Character.getNumericValue(c) : c;
    }

    @Override
    public Token build() {
        return new NumericToken(length, value);
    }
}
